src/librustc_passes/rvalue_promotion.rs

```diff
@@ -412,47 +441,53 @@ fn check_expr<'a, 'tcx>(
                         ty::ImplContainer(_) => {
                             v.handle_const_fn_call(did, node_ty, e.span)
                         }
-                        ty::TraitContainer(_) => v.promotable = false
+                        ty::TraitContainer(_) => Err(NotPromotable),
                     }
                 }
-                _ => v.promotable = false
-            }
+                _ => Err(NotPromotable),
+            };
+            def_result.and(call_result)
```

Comment:
You don't need to do anything like this, you can return early at the first sign of `Err`.

Reply:
All `visit_*` calls need to happen, otherwise if e.g. the first function argument is not promotable, the next one isn't going to be inserted into the promotable map, even if it were promotable

In this case we can early abort right after the for loop though

Reply:
Yea we could move to a custom enum. Do you prefer that? At the beginning I thought the same thing wrt early return, but it's actually pretty easy to screw up, so maybe a custom enum would be better